modules/renter/hostdb/hosttree/hosttree_test.go

```diff
@@ -0,0 +1,565 @@
+		if err != nil {
+			break
+		}
+		node, err := tree.root.nodeAtWeight(types.NewCurrency(randWeight))
+		if err != nil {
+			break
+		}
+		node.remove()
+		delete(tree.hosts, string(node.entry.PublicKey.Key))
+
+		// remove the entry from the hostdb so it won't be selected as a
+		// repeat
+		removedEntries = append(removedEntries, node.entry)
+	}
+	for _, entry := range removedEntries {
+		tree.Insert(entry.HostDBEntry)
+	}
+	return nil
+}
+
+// makeHostDBEntry makes a new host entry with a random public key and the weight
+// provided to `weight`.
+func makeHostDBEntry() modules.HostDBEntry {
+	dbe := modules.HostDBEntry{}
+
+	pk := types.SiaPublicKey{
+		Algorithm: types.SignatureEd25519,
+		Key:       make([]byte, 32),
+	}
+	_, err := io.ReadFull(rand.Reader, pk.Key)
+	if err != nil {
+		panic(err)
+	}
+
+	dbe.AcceptingContracts = true
+	dbe.PublicKey = pk
+
+	return dbe
+}
+
+func TestHostTree(t *testing.T) {
+	tree := New(func(hdbe modules.HostDBEntry) types.Currency {
+		return types.NewCurrency64(20)
+	})
+
+	// Create a bunch of host entries of equal weight.
+	firstInsertions := 64
+	var keys []types.SiaPublicKey
+	for i := 0; i < firstInsertions; i++ {
+		entry := makeHostDBEntry()
+		keys = append(keys, entry.PublicKey)
+		err := tree.Insert(entry)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	err := verifyTree(tree, firstInsertions)
+	if err != nil {
+		t.Error(err)
+	}
+
+	var removed []types.SiaPublicKey
+	// Randomly remove hosts from the tree and check that it is still in order.
+	for _, key := range keys {
+		shouldRemove := func() bool {
+			n, err := rand.Int(rand.Reader, big.NewInt(1))
+			if err != nil {
+				t.Fatal(err)
+			}
+			if n.Cmp(big.NewInt(0)) == 0 {
+				return true
+			}
+			return false
+		}()
+
+		if shouldRemove {
+			err := tree.Remove(key)
+			if err != nil {
+				t.Fatal(err)
+			}
+			removed = append(removed, key)
+		}
+	}
+
+	err = verifyTree(tree, firstInsertions-len(removed))
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Do some more insertions.
+	secondInsertions := 64
+	for i := firstInsertions; i < firstInsertions+secondInsertions; i++ {
+		entry := makeHostDBEntry()
+		tree.Insert(entry)
+	}
+	err = verifyTree(tree, firstInsertions-len(removed)+secondInsertions)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+// Verify that inserting, fetching, deleting, and modifying in parallel from
+// the hosttree does not cause inconsistency.
+func TestHostTreeParallel(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	tree := New(func(dbe modules.HostDBEntry) types.Currency {
+		return types.NewCurrency64(10)
+	})
+
+	// spin up 100 goroutines all randomly inserting, removing, modifying, and
+	// fetching nodes from the tree.
+	var tg siasync.ThreadGroup
+	nthreads := 100
+	nelements := 0
+	var mu sync.Mutex
+	for i := 0; i < nthreads; i++ {
+		go func() {
+			tg.Add()
+			defer tg.Done()
+
+			inserted := make(map[string]modules.HostDBEntry)
```

Comment:
mildly unrealistic because two threads are never working on the same element in the tree at the same time. Given how the tree is implemented, I don't think this is actually an issue, the parallelism should still cause a lot of damage if the implementation is correct.

But generally you want your threads to have data conflicts as well as codepath conflicts when testing.

For this, that might be difficult to do because you'd need to wrap the map in a lock and that might be sufficient to make the threads effectively single threaded. 

It's probably fine as-is, I really don't expect the parallelism code that we added to cause issues.